### PR TITLE
feat: braze ecommerce events onboarding

### PR DESCRIPTION
**Pull request by ItsSudip**
## PR Description

Adds support for Braze's [recommended ecommerce events](https://www.braze.com/docs/user_guide/data/activation/events/recommended_events) in the **web device mode** integration, mirroring the cloud-mode mapping. Gated behind a new opt-in config flag `useEcommerceRecommendedEvents` (default off); when off, the existing legacy custom-event / `logPurchase` path is unchanged.

When enabled, mapped RudderStack ecommerce events are emitted via `globalThis.braze.logCustomEvent(name, properties)` (including `Order Completed`, which maps to `ecommerce.order_placed` rather than `logPurchase`).

**Event mapping (RS → Braze):**

| RudderStack event | Braze event | Notes |
|---|---|---|
| Product Viewed | `ecommerce.product_viewed` | flat, no `products[]` |
| Product Added | `ecommerce.cart_updated` | `action: 'add'`, single-product wrap |
| Product Removed | `ecommerce.cart_updated` | `action: 'remove'` |
| Checkout Started | `ecommerce.checkout_started` | |
| Order Completed | `ecommerce.order_placed` | |
| Order Refunded | `ecommerce.order_refunded` | |
| Order Cancelled | `ecommerce.order_cancelled` | |

`Cart Viewed` / `Cart Updated` are intentionally unmapped and fall through to the legacy path.

**Changes:**
- `ecommerceUtil.js` (new) — event-name resolution, per-event + per-product field mappings, `products[]` construction, unmapped-key → `metadata` pass-through, and `source` derivation (`'web'`, with valid `properties.source` taking precedence).
- `browser.js` — reads the `useEcommerceRecommendedEvents` flag and routes mapped events through the new builder.
- Validation posture matches cloud's "send-anyway": on missing Braze-required fields, emit a single `logger.warn` and still call `logCustomEvent`.
- Empty `products[]` and other empty/null fields are stripped from the payload before sending (the missing-field warning still fires).

## Linear task

[INT-4303](https://linear.app/rudderstack/issue/INT-4303/braze-update-with-recommended-events-in-web-device-mode)

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

* **New Features**
  * Added support for Braze recommended ecommerce events. When enabled, ecommerce events are routed to Braze’s dedicated ecommerce endpoints with correctly normalized sources, mapped fields, and products payload shaping.
* **Bug Fixes**
  * Improved cart, checkout, and order handling, including correct precedence for “order completed,” consistent action derivation, and scrubbing of null/empty product and metadata values.
  * Missing or empty required ecommerce data now triggers warnings while still sending the event; unmapped events use the legacy tracking path.
* **Tests**
  * Added comprehensive Jest coverage for event mapping, payload transformations, and edge cases.
* **Chores**
  * Updated the legacy CDN bundle size limit.

### CODE REVIEW

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+ * 4. Route unmapped event-level keys to `properties.metadata`, and unmapped per-product
+ *    keys to `products[].metadata`.
+ * 5. Emit a single `logger.warn` listing any missing Braze-required fields.
+ *
+ * Never throws on data shape; the warning + the (still-sent) payload is the contract.
+ */
+export const buildEcommerceEventProperties = (message, brazeEvent, action, logger) => {
+  const properties = message.properties || {};
+  const eventMapping = PER_EVENT_MAPPING[brazeEvent];
+  const hasProducts = brazeEvent !== BRAZE_ECOMMERCE_EVENTS.PRODUCT_VIEWED;
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js:478]: Good catch — added `const eventMapping = PER_EVENT_MAPPING[brazeEvent] || [];` so the helpers can't crash on an unknown event, keeping the documented "never throws" contract.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+  if (action) {
+    payload.action = action;
+  }
+
+  // Step 4: route unmapped event-level keys to metadata.
+  const consumedEventKeys = consumedTopLevelKeysForEvent(brazeEvent, eventMapping, hasProducts);
+  const eventMetadata = pickUnmappedKeys(properties, consumedEventKeys);
+  if (Object.keys(eventMetadata).length > 0) {
+    payload.metadata = eventMetadata;
+  }
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Fixed — `action` is now added to the consumed-key set when it's set explicitly, so a caller-provided `properties.action` no longer duplicates into `metadata.action`. Added a regression test.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+const ORDER_PLACED_MAPPING = [
+  m('order_id', 'properties.order_id', true),
+  m('total_value', TOTAL_VALUE_SOURCES, true),
+  m('currency', 'properties.currency', true),
+  m('cart_id', 'properties.cart_id'),
+  m('tax', 'properties.tax'),
+  m('shipping', 'properties.shipping'),
+  m('total_discounts', 'properties.discount'),
+  m('subtotal_value', 'properties.subtotal_value'),
+  m('discounts', 'properties.discounts'),
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Addressed — introduced a shared `TOTAL_DISCOUNTS_SOURCES = ['properties.discount', 'properties.total_discounts']` fallback chain now used by `order_placed`, `order_refunded`, and `order_cancelled`, so all three accept either key. Added a regression test.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+  return consumed;
+};
+
+/**
+ * Build the `products[]` array for the outgoing payload.
+ * - cart_updated: read top-level product fields directly from `properties` into a
+ *   1-element products[]. No per-product metadata — unmapped event-level keys flow
+ *   through the event-level metadata pass.
+ * - other product-bearing events: map each item in `properties.products` and route
+ *   per-product unmapped keys to `products[i].metadata`.
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Updated the comment to reflect that the single-product wrap applies only to `cart_updated` *without* an explicit `products[]`; when an array is present it's mapped like other product-bearing events.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+const ORDER_CANCELLED_MAPPING = [
+  m('order_id', 'properties.order_id', true),
+  m('total_value', TOTAL_VALUE_SOURCES, true),
+  m('currency', 'properties.currency', true),
+  m('cancel_reason', ['properties.cancel_reason', 'properties.reason'], true),
+  m('tax', 'properties.tax'),
+  m('shipping', 'properties.shipping'),
+  m('total_discounts', 'properties.discount'),
+  m('subtotal_value', 'properties.subtotal_value'),
+  m('discounts', 'properties.discounts'),
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Resolved with the same shared `TOTAL_DISCOUNTS_SOURCES` fallback chain as `order_placed` — see the related thread above.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+/**
+ * Build the `properties` object for a Braze recommended ecommerce event.
+ *
+ * Algorithm:
+ * 1. Run `constructPayload` against the message event-level mapping (never throws —
+ *    send-anyway is enforced via the validation warning instead).
+ * 2. For events with a `products[]`, build the array (single-product wrap for
+ *    `cart_updated`, iterate `properties.products` otherwise).
+ * 3. Set `source` via `deriveSource` and `action` when present.
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Updated the algorithm doc to note the single-product wrap is only for `cart_updated` without an explicit `products[]`.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+  if (hasProducts) {
+    consumed.add('products');
+  }
+
+  // cart_updated uses top-level product fields as the single wrapped product;
+  // mark those keys as consumed so they don't duplicate into event-level metadata.
+  if (brazeEvent === BRAZE_ECOMMERCE_EVENTS.CART_UPDATED && hasProducts) {
+    consumedKeysFromMapping(ECOMMERCE_PRODUCT_MAPPING).forEach(key => consumed.add(key));
+  }
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Good find — fixed. `consumedTopLevelKeysForEvent` now takes `properties` and only consumes the top-level product fields for `cart_updated` when there's no explicit `products[]`. With an array present, the top-level fields flow through to `metadata` as expected. Added a regression test asserting top-level fields survive into `metadata` when a `products[]` array coexists.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+  const consumedKeys = consumedKeysFromMapping(ECOMMERCE_PRODUCT_MAPPING);
+  return rawProducts.map(raw => {
+    const item = raw && typeof raw === 'object' ? raw : {};
+    const product = constructPayload(item, ECOMMERCE_PRODUCT_MAPPING) || {};
+    const productMetadata = pickUnmappedKeys(item, consumedKeys);
+    if (Object.keys(productMetadata).length > 0) {
+      product.metadata = productMetadata;
+    }
+    return product;
+  });
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Good catch — fixed all three. The single-product wrap now scrubs the constructed product and returns `[]` (which gets stripped from the payload entirely) instead of `[{}]` when there are no mapped fields; the `products[]` path now `.filter()`s out fully-empty product objects; and per-product `metadata` is scrubbed of null/empty values via `pickUnmappedKeys`. Added a regression test asserting no degenerate empty product is emitted.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+    hasProducts,
+    properties,
+  );
+  if (action) {
+    consumedEventKeys.add('action');
+  }
+  const eventMetadata = pickUnmappedKeys(properties, consumedEventKeys);
+  if (Object.keys(eventMetadata).length > 0) {
+    payload.metadata = eventMetadata;
+  }
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js:508]: Fixed — `pickUnmappedKeys` now runs the result through `removeUndefinedAndNullAndEmptyValues`, which is the single chokepoint for both event-level and per-product metadata, so null/empty values are stripped before the metadata is attached. Added a regression test asserting null/empty values don't survive into either `metadata` or `products[].metadata`.

File: packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js [hunk 10]
```diff
@@ -1077,6 +1077,361 @@ describe('track', () => {
...
+  });
+
+  it('does not emit an empty product object when cart_updated has no product fields', () => {
+    const braze = buildBraze();
+    trackEvent(braze, 'Cart Updated', { cart_id: 'c1' });
+
+    // no degenerate `[{}]` — the empty products array is scrubbed off entirely
+    const props = window.braze.logCustomEvent.mock.calls[0][1];
+    expect(props.products).toBeUndefined();
+  });
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js:1242]: Good catch — you're right, `Cart Updated` isn't in the mapping table (only `Product Added`/`Product Removed` map to `ecommerce.cart_updated`), so the test was falling through to the legacy path and passing vacuously. Switched it to `Product Added` with no product fields and added an assertion that the emitted event name is `ecommerce.cart_updated`, so it now actually covers the "no degenerate `[{}]`" scenario through `buildProductsArray`.

File: packages/analytics-js-integrations/src/integrations/Braze/browser.js [hunk 11]
```diff
@@ -357,7 +359,19 @@ class Braze {
...
+      if (ecommerceMapping) {
+        const { brazeEvent, action } = ecommerceMapping;
+        const ecommerceProperties = buildEcommerceEventProperties(
+          rudderElement.message,
+          brazeEvent,
+          action,
+          logger,
+        );
+        globalThis.braze.logCustomEvent(brazeEvent, ecommerceProperties);
+      } else if (eventName.toLowerCase() === 'order completed') {
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/browser.js:374]: We can ignore the old code path.

File: packages/analytics-js-integrations/.size-limit.mjs [hunk 12]
```diff
@@ -6,7 +6,7 @@ export default [
   {
     name: 'All Integrations - Legacy - CDN',
     path: 'dist/cdn/legacy/js-integrations/*.min.js',
-    limit: '97 KiB',
+    limit: '98.5 KiB',
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/.size-limit.mjs:9]: I have discussed with the codeowners regarding this.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+    brazeEvent,
+    eventMapping,
+    hasProducts,
+    properties,
+  );
+  if (action) {
+    consumedEventKeys.add('action');
+  }
+  const eventMetadata = pickUnmappedKeys(properties, consumedEventKeys);
+  if (Object.keys(eventMetadata).length > 0) {
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js:506]: We can send these values to braze. They are expecting it in metadata.

**Review by achettyiitr — DISMISSED** [review]

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 14]
```diff
@@ -0,0 +1,522 @@
...
+    );
+    return Object.keys(product).length > 0 ? [product] : [];
+  }
+
+  const rawProducts = Array.isArray(properties.products) ? properties.products : [];
+  const consumedKeys = consumedKeysFromMapping(ECOMMERCE_PRODUCT_MAPPING);
+  return rawProducts
+    .map(raw => {
+      const item = raw && typeof raw === 'object' ? raw : {};
+      const product = removeUndefinedAndNullAndEmptyValues(
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js:354]: It is fine. Our customers generally don't send products array like this.

File: packages/analytics-js-integrations/src/integrations/Braze/browser.js [hunk 15]
```diff
@@ -37,6 +38,7 @@ class Braze {
       }
     }
     this.endPoint = '';
+    this.useEcommerceRecommendedEvents = config.useEcommerceRecommendedEvents || false;
     this.isHybridModeEnabled = config.connectionMode === 'hybrid';
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/browser.js:42]: We are using useEcommerceRecommendedEvents only. Updated the description.

File: packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js [hunk 16]
```diff
@@ -1077,6 +1077,519 @@ describe('track', () => {
...
 
+describe('track - recommended ecommerce events', () => {
+  const baseConfig = {
+    appKey: 'APP_KEY',
+    trackAnonymousUser: true,
+    enableBrazeLogging: false,
+    dataCenter: 'US-03',
+    allowUserSuppliedJavascript: false,
+    useEcommerceRecommendedEvents: true,
+  };
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js:1088]: We are using useEcommerceRecommendedEvents only so we can resolve the comment.

File: packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js [hunk 17]
```diff
@@ -1077,6 +1077,519 @@ describe('track', () => {
...
+
+    expect(globalThis.braze.logCustomEvent).toHaveBeenCalledWith('Cart Updated', {
+      cart_id: 'c1',
+      value: 10,
+    });
+  });
+
+  it('falls through to legacy purchase path when the flag is off', () => {
+    const braze = buildBraze({ useEcommerceRecommendedEvents: false });
+    trackEvent(braze, 'order completed', {
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js:1572]: Updated the documentation we are using useEcommerceRecommendedEvents only.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 18]
```diff
@@ -0,0 +1,522 @@
...
+const matchesType = (value, type) => {
+  switch (type) {
+    case FIELD_TYPE.STRING:
+      return typeof value === 'string';
+    case FIELD_TYPE.INTEGER:
+      // `value % 1 === 0` is true only for whole numbers; NaN/Infinity yield NaN and fail.
+      return typeof value === 'number' && value % 1 === 0;
+    case FIELD_TYPE.FLOAT:
+      return typeof value === 'number' && !Number.isNaN(value);
+    case FIELD_TYPE.STRING_ARRAY:
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js:237]: Addressed

File: packages/analytics-js-integrations/src/integrations/Braze/browser.js [hunk 19]
```diff
@@ -37,6 +38,7 @@ class Braze {
       }
     }
     this.endPoint = '';
+    this.useEcommerceRecommendedEvents = config.useEcommerceRecommendedEvents || false;
```
Comment (saikumarrs) [inline packages/analytics-js-integrations/src/integrations/Braze/browser.js]: Fail-safe.
```suggestion
    this.useEcommerceRecommendedEvents = config.useEcommerceRecommendedEvents === true;
```

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 20]
```diff
@@ -0,0 +1,542 @@
...
+// Per-event field mappings (mirror of the cloud `Braze*Config.json` files).
+// Each entry is built via `m(destKey, sourceKeys, required, type)`:
+//   - `destKey`/`sourceKeys` are the `constructPayload` contract.
+//   - `req` flags Braze-required fields (consumed by collectMissingRequiredFields).
+//   - `type` is the Braze-expected type (default String); drives coercion + the
+//     type-mismatch warning.
+// `sourceKeys` arrays are ordered fallback chains (first resolved value wins).
+// ---------------------------------------------------------------------------
+
+const m = (destKey, sourceKeys, req = false, type = FIELD_TYPE.STRING) => ({
```
Comment (saikumarrs) [inline packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: `m` for a method name is not readable at all.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 21]
```diff
@@ -0,0 +1,542 @@
...
+export const BRAZE_ECOMMERCE_EVENTS = {
+  PRODUCT_VIEWED: 'ecommerce.product_viewed',
+  CART_UPDATED: 'ecommerce.cart_updated',
+  CHECKOUT_STARTED: 'ecommerce.checkout_started',
+  ORDER_PLACED: 'ecommerce.order_placed',
+  ORDER_REFUNDED: 'ecommerce.order_refunded',
+  ORDER_CANCELLED: 'ecommerce.order_cancelled',
+};
+
+const BRAZE_SOURCE_VALUES = ['web', 'ios', 'android'];
```
Comment (saikumarrs) [inline packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: How are source values other than web applicable for a device mode integration?

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 22]
```diff
@@ -0,0 +1,542 @@
...
+export const getEcommerceMapping = eventName => {
+  if (typeof eventName !== 'string') {
+    return undefined;
+  }
+  return EVENT_NAME_TO_BRAZE[eventName.trim().toLowerCase()];
+};
+
+/**
+ * Derive the Braze `source` field. On the web SDK this resolves to `'web'`, unless the
+ * event carries an explicit, valid `properties.source` (`web`/`ios`/`android`), which
```
Comment (saikumarrs) [inline packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: I'm not sure if it's relevant for anyone to use a web SDK to send data for other source types.

**Review by saikumarrs — COMMENTED** [review]
I haven't reviewed any of the functionality added but focussed mainly on structural aspects.

File: packages/analytics-js-integrations/src/integrations/Braze/browser.js (hunk 19 above)
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/browser.js]: Applied. Agreed — `|| false` would enable the feature on any truthy value (e.g. a stringified `"false"` coming from config), whereas `=== true` fails safe to off. It also matches the `usePlatformSpecificApiKeys === true` convention right above it.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js (hunk 20 above)
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Renamed to `mapField` (declaration + all 51 call sites), e.g. `mapField('price', 'properties.price', true, FIELD_TYPE.FLOAT)`. No bundle cost — it's a module-local const, so it gets minified regardless.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js (hunk 21 above)
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: You're right — removed. This is the web device-mode integration, so `source` is now hardcoded to `web` via a single `BRAZE_WEB_SOURCE` const; `BRAZE_SOURCE_VALUES` and `deriveSource` are deleted entirely. `ios`/`android` stay in the mobile SDKs, where they actually apply.

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js (hunk 22 above)
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js]: Agreed, and fixed together with the `BRAZE_SOURCE_VALUES` thread — the `properties.source` override is gone. `source` is always `web` here, and a caller-supplied `properties.source` is now ignored *and* doesn't leak into `metadata`. Added a regression test asserting both.

File: packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js [hunk 23]
```diff
@@ -1077,6 +1077,396 @@ describe('track', () => {
...
+    trackEvent(braze, 'Product Viewed', {
+      product_id: 'p1',
+      name: 'Game',
+      variant: 'v1',
+      price: 15.99,
+      currency: 'USD',
+      source: 'ios ',
+    });
+
+    expect(window.braze.logCustomEvent.mock.calls[0][1].source).toBe('ios');
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/__tests__/integrations/Braze/browser.test.js]: Done — converted the assertions in the `track - recommended ecommerce events` block to `globalThis.braze` (24 occurrences), matching how the integration dispatches and keeping them worker-compatible. Left the pre-existing tests outside this PR's scope on `window` to keep the diff focused.

**Review by saikumarrs — APPROVED** [review]

File: packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js [hunk 24]
```diff
@@ -0,0 +1,527 @@
...
+  // Step 1: event-level field mapping, with each mapped value coerced to its Braze type.
+  const payload = coerceMappedFields(constructPayload(message, eventMapping) || {}, eventMapping);
+
+  // Step 2: products[] (skipped for product_viewed — flat, single-product event).
+  if (hasProducts) {
+    payload.products = buildProductsArray(properties, brazeEvent);
+  }
+
+  // Step 3: source + action.
+  payload.source = BRAZE_WEB_SOURCE;
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply packages/analytics-js-integrations/src/integrations/Braze/ecommerceUtil.js:489]: We can ignore this

**Review by achettyiitr — APPROVED** [review]
